A class that implements matrix scaling algorithms sequentially

PolvereUnipi opened the pull request:
This PR intends to add the MatrixScaler class to deal. The class implements algorithm for matrix scaling through the action of diagonal matrices. I'd like to allow the possibility of eventually adding new scaling algorithms in the future. 

This PR contains only the sequential versions of the algorithms. The aim is to showcase the class and start a discussion on the class and its implementation in case a different design is needed or changes are requested. Some simple test have also been added.

I am also working on the MPI implementation of the algorithms. The MPI implementation will be added on a follow up PR.   

The scaling algorithm implemented are Sinkhorn-Knopp scaling https://doi.org/10.1137/060659624 and symmetry preserving scaling algorithm https://doi.org/10.1137/110825753 . 

@fdrmrc @luca-heltai

PolvereUnipi commented:
@luca-heltai 

Thank you! 

I've addressed all your comments, fixed the file names of the tests and squashed all the commits.

PolvereUnipi commented:
@luca-heltai 
I fixed the typo. I also re-introduced the default on the destructor since tidy check was failing due to missing default. 

I should also have fixed the failing Jenkins: MPI and Serial tests.

luca-heltai commented:
Yes please. Also, would you mind adding a small test for block matrices and block vectors? Thanks!

luca-heltai commented:
@PolvereUnipi is my PhD student, so I'll let someone else chime in with the approval.

bangerth commented on include/deal.II/lac/block_matrix_base.h:
```diff
@@ -1350,8 +1350,8 @@ namespace BlockMatrixIterators
-        this->row_block = numbers::invalid_size_type;
-        this->col_block = numbers::invalid_size_type;
+        this->row_block = numbers::invalid_unsigned_int;
+        this->col_block = numbers::invalid_unsigned_int;
```
For the future: the changes in this file are logically unrelated to what you do in this pull request. Put unrelated changes into separate PRs. That makes it much easier to review and get things merged!

bangerth commented on include/deal.II/lac/matrix_scaling.h:
```diff
@@ -0,0 +1,283 @@
+ * scaling algorithms scale the matrix in place by the action of two diagonal
+ * matrices: one acting on the rows and one acting on the columns. The diagonals
+ * of the two scaling matrices applied on the last scaled matrix are stored in
+ * the class as vectors. The algorithms implemented here are the Sinkhorn-Knopp
```
I don't quite understand what you mean by "the last scaled matrix", but perhaps that's not important: I would take that whole sentence out. What you want to do is document *what* the class does, not how it does it. From a user's perspective, it is not important how exactly these diagonal matrices are actually stored.

bangerth commented on include/deal.II/lac/matrix_scaling.h:
```diff
@@ -0,0 +1,283 @@
+ * algorithm described in this <a
+ * href="https://doi.org/10.1137/060659624">article</a> and the scaling
+ * algorithm that preserves symmetry described in this <a
+ * href="https://doi.org/10.1137/110825753">article</a>.
```
I think a better approach would be if you added these references to `doc/doxygen/references.bib` and then cited them here via `@cite`.

bangerth commented on include/deal.II/lac/matrix_scaling.h:
```diff
@@ -0,0 +1,283 @@
+   */
+  template <class VectorType>
+  void
+  scale_system_solution(VectorType &sol) const;
```
I had to think about this interface for a while, and did not quite make my peace with it :-)

My approach to determining whether something is a good interface is usually to just look at the verbs used in function names. Here, you have a function that says `scale_matrix()`, which one would ordinarily expect to *scale the matrix*, i.e. *operate on its argument*. For example, the `scale_system_solution()` works this way, and consequently it is a `const` function. But the other two `scale_*()` functions are not `const`, and that's because they *also* do something else, namely *analyze* the matrix and store the result of this analysis. 

I have not looked at the implementation yet. If the analysis algorithm can be run without actually changing the matrix, then I think a better interface would be something like this:
```
  void analyze_matrix(const Matrix &m);
  void scale_matrix(Matrix &m) const;                     // can now be const
  void scale_linear_system(Matrix &m, Vector &v) const;  // same here
  void scale_system_solution (Vector &v) const;     // has always been const
```
On the other hand, if the analysis algorithm really requires modifying the matrix, then splitting the tasks as above would require creating a copy of the matrix. We'd like to avoid that, and in that case perhaps a better choice of function names would be this:
```
  void analyze_and_scale_matrix(Matrix &m);
  void analyze_and_scale_linear_system(Matrix &m, Vector &v);
  void scale_system_solution (Vector &v) const;     // has always been const
```
That's a little bit more work, but I think it more clearly explains what these functions do. The word "analyze" expresses that we do something *on* the matrix without changing it, and this also makes it clear that the results of this analysis need to be stored somewhere so that we can use the analysis later on.

bangerth replied:
The analysis part is the computation of the scaling vectors. My comment about some of the functions being `const` whereas others are not was about whether one would *expect* functions with these names to be `const` or not. I understand that *for the implementation you have*, some of them need to be non-`const`, but that's not the right approach to software design: Start with what you *want* things to be, and then find an implementation that achieves that. 

Either way, I think we understand each other :-)

bangerth commented on include/deal.II/lac/matrix_scaling.h:
```diff
@@ -0,0 +1,283 @@
+   */
+  template <class Matrix>
+  void
+  sk_scaling(Matrix &matrix, const unsigned int nsteps);
```
These 3 functions are missing a verb in their name. Perhaps `do_*_scaling()`?

bangerth commented on source/lac/matrix_scaling.cc:
```diff
@@ -0,0 +1,440 @@
+    {
+      (void)matrix;
+      Assert(false, ExcNotImplemented());
+    }
```
This is in the .cc file, so the function is only instantiated for matrix types *you* choose. We don't need to expect users to instantiate this class, so you might as well break compilation for all other matrix types by putting the following at the top of the function.
```
  static_assert (is_supported_matrix<Matrix>());
```
Then you can get rid of the `if constexpr`.

bangerth commented on source/lac/matrix_scaling.cc:
```diff
@@ -0,0 +1,440 @@
+                {
+                  for (auto it = matrix.begin(row); it != matrix.end(row); ++it)
+                    {
+                      // it->value() /= col_norms[it->column()];
```
and here

bangerth commented on source/lac/matrix_scaling.cc:
```diff
@@ -0,0 +1,440 @@
+                {
+                  for (auto it = matrix.begin(row); it != matrix.end(row); ++it)
+                    {
+                      // it->value() /= row_norms[row];
```
and here

bangerth commented on source/lac/matrix_scaling.cc:
```diff
@@ -0,0 +1,440 @@
+        }
+        break;
+      case MatrixScaling::AdditionalData::SKParameters::NormType::l_infinity:
+        {
```
It would be easier to read with the `break` statement inside the curly braces:
```suggestion
          break;
        }

            case MatrixScaling::AdditionalData::SKParameters::NormType::l_infinity:
        {
```
I also gave it an extra line to make it easier to read.

PolvereUnipi commented:
@bangerth 

I addressed your comments and pushed some of the suggested changes on the MPI branch. The MPI PR #18911 is more polished than this one and also some issues that were related to the sequential part and that you pointed out were addressed there. 

Sorry for the mess, i think it's easier for me to address everything there. I will close this PR  since everything that is here is also on #18911.
